h/api/search/index.py

```diff
@@ -69,3 +73,183 @@ def delete(es, annotation_id):
+            for a in batch:
+                yield a
+
+    def _stream_filtered_annotations(self, annotation_ids):
+        annotations = self._eager_loaded_query(). \
+            execution_options(stream_results=True). \
+            filter(models.Annotation.id.in_(annotation_ids)).all()
+
+        for a in annotations:
+            yield a
+
+    def _eager_loaded_query(self):
+        return self.session.query(models.Annotation).options(
+            subqueryload(models.Annotation.document).subqueryload(models.Document.document_uris),
+            subqueryload(models.Annotation.document).subqueryload(models.Document.meta)
+        )
+
+
+class BatchDeleter(object):
+    """
+    A convenience class for removing all annotations that are deleted from the
+    database from the search index.
+    """
+
+    def __init__(self, session, es_client):
+        self.session = session
+        self.es_client = es_client
+
+    def delete_all(self):
+        """Remove all deleted annotations, and retry failed delete operations once."""
+        ids = self.deleted_annotation_ids()
+        for _ in range(2):
+            if not ids:
+                break
+
+            ids = self.delete(ids)
+            log.debug(
+                'Failed to delete {} annotations, might retry'.format(len(ids)))
+
+    def deleted_annotation_ids(self):
+        """
+        Get a list of deleted annotation ids by comparing all ids in the search
+        index and comparing them to all ids in the database.
+
+        :returns: a set of deleted annotation ids
+        :rtype: set
+        """
+        ids = set()
+        for batch in self._batch_iter(2000, self._es_scan()):
+            ids.update({a['_id'] for a in batch})
+
+        ids.difference_update(self._pg_ids())
+        return ids
+
+    def delete(self, annotation_ids):
+        """
+        Delete annotations from the search index.
+
+        :param annotation_ids: a list of ids to delete.
+        :type annotation_ids: collection
+
+        :returns: a set of errored ids
+        :rtype: set
+        """
+        if not annotation_ids:
+            return set()
+
+        deleting = es_helpers.streaming_bulk(self.es_client.conn, annotation_ids,
+                                             chunk_size=100,
+                                             expand_action_callback=self._present,
+                                             raise_on_error=False)
+        errored = set()
+        for ok, item in deleting:
+            if not ok and item['delete']['status'] != 404:
+                errored.add(item['delete']['_id'])
+        return errored
+
+    def _present(self, id_):
+        action = {'delete': {'_index': self.es_client.index,
+                             '_type': self.es_client.t.annotation,
+                             '_id': id_}}
+        return (action, None)
+
+    def _pg_ids(self):
+        # This is using a Postgres cursor for better query performance.
```

Comment:
It is now returning a SQLAlchemy query and we use the `batch_iter` function in `deleted_annotation_ids`, very similar to the way we load ElasticSearch annotation ids. :+1: